slides/advanced-01-tuning-hyperparameters.qmd

```diff
@@ -0,0 +1,697 @@
+
+# Activation function in neural networks?
+
+Sigmoidal functions, ReLu, etc.
+
+::: fragment
+Yes, it is a tuning parameter.
+✅
+:::
+
+# Number of PCA components to retain?
+
+::: fragment
+Yes, it is a tuning parameter.
+✅
+:::
+
+# Bayesian priors for model parameters?
+
+::: fragment
+Hmmmm, probably not.
+These are based on prior belief.
+❌
+:::
+
+# Covariance/correlation matrix structure in mixed models?
+
+::: fragment
+Yes, but it is unlikely to affect performance.
+:::
+
+::: fragment
+It will impact inference though.
+🤔
+:::
+
+
+
+# Is the random seed a tuning parameter?
+
+::: fragment
+Nope. It is not. 
+❌
+:::
+
+## Optimize tuning parameters
+
+- Try different values and measure their performance.
+
+. . .
+
+- Find good values for these parameters.
+
+. . .
+
+- Once the value(s) of the parameter(s) are determined, a model can be finalized by fitting the model to the entire training set.
+
+
+## Boosted Trees
+
+These are popular ensemble methods that build a _sequence_ of tree models. 
+
+<br>
+
+Each tree uses the results of the previous tree to better predict samples, especially those that have been poorly predicted. 
+
+<br>
+
+Each tree in the ensemble is saved and new samples are predicted using a weighted average of the votes of each tree in the ensemble. 
+
+<br>
+
+We'll focus on the popular lightgbm implementation. 
+
+## Boosted Tree Tuning Parameters
+
+* The number of predictors randomly sampled at each split (aka `mtry` in $[1, ncol(x)]$ or $(0, 1]$).
+* The number of trees (`trees` in $[1, \infty]$, but usually up to thousands)
+* The number of samples needed to further split (`min_n` in $[1, n]$).
+* The rate that each tree adapts form previous iterations (`learn_rate` in $(0, \infty]$, usual maximum is 0.1).
+* The number of iterations of boosting where _no improvement_ was shown (`stop_iter`)
+
+## Boosted Tree Tuning Parameters
+
+TBH it is usually not difficult to optimize these models. 
+
+<br>
+
+Often, there are multiple _candidate_ tuning parameter combinations that have very good results. 
+
+<br>
+
+To demonstrate simple concepts, we'll look at optimizing the number of trees in the ensemble (between 1 and 100) and the learning rate ($10^{-5}$ to $10^{-1}$).
+
+## Boosted Tree Tuning Parameters `r hexes(c("workflows", "parsnip", "bonsai"))`
+
+We'll need to load the bonsai package. This has the information needed to use lightgbm
+
+```{r}
+#| label: boot-spec
+
+library(bonsai)
+lgbm_spec <- 
+  boost_tree(trees = tune(), learn_rate = tune()) %>% 
+  set_mode("regression") %>% 
+  set_engine("lightgbm")
+
+lgbm_wflow <- workflow(hotel_rec, lgbm_spec)
+```
+
+
+## Optimize tuning parameters
+
+The main two strategies for optimization are:
+
+. . .
+
+- **Grid search** 💠 which tests a pre-defined set of candidate values
+
+- **Iterative search** 🌀 which suggests/estimates new values of candidate parameters to evaluate
+
+We'll focus on grid search for a while. 
```

Comment:
```suggestion
```

Since you discuss both strategies right after this but then have a "Grid Search" title slide, I'd say this is better left out.